lib/index.js

```diff
@@ -36,143 +68,112 @@ module.exports = {
-            debug('routing %s@%s for %s', routePath, route.version, route.name);
-            if (route.method !== 'opts') {
-              server.opts(routePath,
-                cors.preflight({ extraHeaders: route.expectedHeaders }));
-            }
-            server[route.method](
-              { path: routePath, version: route.version },
-              cors.headers(),
-              route.handler
-            );
+            expectedHeaders: actionExpectedHeaders,
+            handler: routeHandler(resolved, r, a, require(controllerPath), options.context)
           });
         });
-        return q.resolve();
-      })
-      .then(function() {
-        server.on('after', function(req, res, route, err) {
-          logger(req, res, route, err);
-        });
-        server.on('NotFound', function(req, res, err, cb) {
-          res.send(404, { type: 'NotFound', message: err.message });
-          cb();
-        });
-        server.on('MethodNotAllowed', function(req, res, err, cb) {
-          res.send(405, { type: 'MethodNotAllowed', message: err.message });
-          cb();
-        });
-        server.on('VersionNotAllowed', function(req, res, err, cb) {
-          res.send(400, { type: 'VersionNotAllowed', message: err.message });
-          cb();
-        });
-        server.on('InternalServer', function(req, res, err, cb) {
-          res.send(500, { type: 'ServerError', message: err.message });
-          cb();
-        });
-        server.on('uncaughtException', function(req, res, route, err) {
-          logger(req, res, route, err);
-          if (!res.headersSent) {
-            res.send(err.statusCode || 500, { type: err.type || 'Error', message: err.message });
-          }
+      });
+      routes[api.basePath] = [{
+        name: 'index',
+        method: 'get',
+        version: api.info.version,
+        handler: indexRouteHandler(api)
+      }];
+    });
+
+
+    const corsMap = {};
+    keys(routes).sort().reverse().forEach((routePath) => {
+      corsMap[routePath] = { shouldSet: true, expectedHeaders: new Set(), allowedMethods: new Set(['OPTIONS']) };
+      routes[routePath].forEach((action) => {
+        if (action.method === 'opts') { corsMap[routePath].shouldSet = false; }
+        corsMap[routePath].allowedMethods.add(action.method.toUpperCase());
+        (action.expectedHeaders || []).forEach((header) => {
+          corsMap[routePath].expectedHeaders.add(header);
         });
 
-        server.listen(
-          config.get('port'),
-          config.get('host'),
-          function() {
-            debug('%s listening at %s', server.name, server.url);
-          }
+        debug('routing %s@%s for %s', routePath, action.version, action.name);
+        server[action.method](
+          { path: routePath, version: action.version },
+          cors.headers(),
+          action.handler
         );
+      });
+    });
 
-        return server;
+    keys(corsMap).forEach((route) => {
```

Comment:
should this be `sort()`ed and `reverse()`d too? I don't really understand why order matters for setting actions/headers